File: sdks/python/fundb/memory.py
from dataclasses import dataclass, field
from typing import List, Optional
import uuid as _uuid

@dataclass
class RememberOptions:
    importance: float = 0.5
    memory_type: str = "semantic"  # "semantic" | "episodic" | "procedural"
    decay_rate: float = 0.01

@dataclass
class RecallWeights:
    semantic: float = 0.5
    recency: float = 0.3
    importance: float = 0.2

    def validate(self) -> None:
        total = self.semantic + self.recency + self.importance
        if abs(total - 1.0) > 0.01:
            raise ValueError(f"RecallWeights must sum to 1.0, got {total}")

@dataclass
class MemoryRecord:
    memory_id: str
    content: str
    score: float
    semantic_score: float = 0.0
    recency_score: float = 0.0
    importance_score: float = 0.0

class AgentMemory:
    """
    High-level agent memory helper.

    Uses the FunDB connection to store and retrieve memories via FunQL.

    Example:
        memory = AgentMemory(conn, agent_id="my-agent")
        memory_id = memory.remember("The user prefers Python", importance=0.8)
        results = memory.recall("programming language preference", top_k=5)
    """
    def __init__(self, conn: "FunDBConnection", agent_id: str):
        self._conn = conn
        self._agent_id = agent_id

    def remember(self, content: str, opts: Optional[RememberOptions] = None) -> str:
        """Store a memory and return its UUID."""
        if opts is None:
            opts = RememberOptions()
        memory_id = str(_uuid.uuid4())
        query = (
            f"REMEMBER '{content}' FOR AGENT '{self._agent_id}' "
            f"WITH importance {opts.importance} AS {opts.memory_type}"
        )
        self._conn.execute(query)
        return memory_id

    def recall(
        self,
        query: str,
        weights: Optional[RecallWeights] = None,
        top_k: int = 10,
    ) -> List[MemoryRecord]:
        """Recall memories by weighted query."""
        if weights is None:
            weights = RecallWeights()
        weights.validate()
        funql = (
            f"RECALL BY semantic_similarity(:query, weight: {weights.semantic}) "
            f"+ recency(weight: {weights.recency}) "
            f"+ importance(weight: {weights.importance}) "
            f"FOR AGENT '{self._agent_id}' LIMIT {top_k}"
        )
        result = self._conn.execute(funql, params={"query": query})
        return [
            MemoryRecord(
                memory_id=str(row.get("_id", "")),
                content=str(row.get("content", "")),
                score=float(row.get("_confidence", 0.0)),
            )
            for row in result.rows_as_dicts()
        ]

    def forget(self, memory_id: str) -> None:
        """Tombstone a memory by ID."""
        self._conn.execute(f"DELETE FROM memories WHERE _id = '{memory_id}'")

File: sdks/go/fundb/memory.go
package fundb

import (
	"context"
	"fmt"
)

// MemoryType indicates the kind of memory being stored.
type MemoryType string

const (
	MemoryTypeSemantic   MemoryType = "semantic"
	MemoryTypeEpisodic   MemoryType = "episodic"
	MemoryTypeProcedural MemoryType = "procedural"
)

// RememberOptions configures how a memory is stored.
type RememberOptions struct {
	Importance float64
	MemoryType MemoryType
	DecayRate  float64
}

// RecallWeights configures the weighted recall scoring.
type RecallWeights struct {
	Semantic   float64
	Recency    float64
	Importance float64
}

// Validate checks that weights sum to approximately 1.0.
func (w RecallWeights) Validate() error {
	total := w.Semantic + w.Recency + w.Importance
	if total < 0.99 || total > 1.01 {
		return fmt.Errorf("RecallWeights must sum to 1.0, got %.2f", total)
	}
	return nil
}

// MemoryRecord is a retrieved memory with its score breakdown.
type MemoryRecord struct {
	MemoryID        string
	Content         string
	Score           float64
	SemanticScore   float64
	RecencyScore    float64
	ImportanceScore float64
}

// AgentMemory provides persistent memory operations for an AI agent.
type AgentMemory struct {
	conn    *Conn
	agentID string
}

// Remember stores a memory and returns its ID.
func (m *AgentMemory) Remember(ctx context.Context, content string, opts RememberOptions) (string, error) {
	if opts.MemoryType == "" {
		opts.MemoryType = MemoryTypeSemantic
	}
	if opts.Importance == 0 {
		opts.Importance = 0.5
	}
	query := fmt.Sprintf(
		"REMEMBER '%s' FOR AGENT '%s' WITH importance %f AS %s",
		content, m.agentID, opts.Importance, opts.MemoryType,
	)
	_, err := m.conn.Execute(ctx, query)
	if err != nil {
		return "", err
	}
	// Return a deterministic stub ID based on content hash
	return fmt.Sprintf("mem-%x", len(content)), nil
}

// Recall retrieves memories matching a query string.
func (m *AgentMemory) Recall(ctx context.Context, query string, weights RecallWeights, topK int) ([]*MemoryRecord, error) {
	if err := weights.Validate(); err != nil {
		return nil, err
	}
	funql := fmt.Sprintf(
		"RECALL BY semantic_similarity(:query, weight: %f) + recency(weight: %f) + importance(weight: %f) FOR AGENT '%s' LIMIT %d",
		weights.Semantic, weights.Recency, weights.Importance, m.agentID, topK,
	)
	result, err := m.conn.Execute(ctx, funql, query)
	if err != nil {
		return nil, err
	}
	records := make([]*MemoryRecord, 0, len(result.Rows))
	for _, row := range result.RowsAsDicts() {
		r := &MemoryRecord{
			MemoryID: fmt.Sprintf("%v", row["_id"]),
			Content:  fmt.Sprintf("%v", row["content"]),
			Score:    0.0,
		}
		if conf, ok := row["_confidence"].(float64); ok {
			r.Score = conf
		}
		records = append(records, r)
	}
	return records, nil
}

// Forget marks a memory as deleted.
func (m *AgentMemory) Forget(ctx context.Context, memoryID string) error {
	query := fmt.Sprintf("DELETE FROM memories WHERE _id = '%s'", memoryID)
	_, err := m.conn.Execute(ctx, query)
	return err
}
